src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -0,0 +1,72 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql.Util
+{
+    /// <summary>
+    /// Internal struct wrapping a <see cref="CancellationTokenSource"/> for timeouts.
+    /// </summary>
+    /// <remarks>
+    /// Since there's no way to reset a <see cref="CancellationTokenSource"/> once it was cancelled,
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
+        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+
+        public TimeoutCancellationTokenSourceWrapper() => Timeout = System.Threading.Timeout.InfiniteTimeSpan;
+
+        public TimeoutCancellationTokenSourceWrapper(TimeSpan timeout) => Timeout = timeout;
+
+        /// <summary>
+        /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// and make sure that it hasn't been cancelled yet
+        /// </summary>
+        public CancellationToken Start()
+        {
+            _timeoutCts.CancelAfter(Timeout);
+            if (_timeoutCts.IsCancellationRequested)
+            {
+                _timeoutCts.Dispose();
+                _timeoutCts = new CancellationTokenSource(Timeout);
+            }
+
+            return _timeoutCts.Token;
+        }
+
+        /// <summary>
+        /// Reset the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// </summary>
+        public void Restart() => _timeoutCts.CancelAfter(Timeout);
```

Comment:
What happens if _timeoutCts expired in the meantime? This looks unsafe...

Reply:
The Idea is exactly to restart the timeout *without* changing `_timeoutCts` so that we can detect the fact that it had already timed out in the next iteration.
In other words restart the timer without resetting `_timeoutCts`.

Reply:
Ah, I see how it's used now, sorry.

I'd at least consider renaming this because it's not at all clear :sweat_smile: Maybe RestartUnlessExpired?

Reply:
Another (probably better) option is simply to not do this pattern, and simply do ThrowIfCancellationRequested on the token (in ReadBuffer.EnsureLong), and then Stop/Start it (possibly via a method called Restart) - there's no real reason for this pattern of restarting only if it hasn't expired.

Reply:
I still see a use case for this method.
It is very slim which is good for tight loops in the hot path and not totally insane if you know what you're doing.
I've named it `RestartTimeoutWithoutReset()` now and extended the documentation comment.

Reply:
Is there an actual hot path at the moment where we think it's needed?

Reply:
Well, we're talking about the [loop where we're reading from the stream in `NpgsqlReadBuffer`](https://github.com/npgsql/npgsql/blob/3c70e78e60511d6b66a01769a75d86caabdc162b/src/Npgsql/NpgsqlReadBuffer.cs#L199) which I'd consider hot path.